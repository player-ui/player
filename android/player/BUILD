load(":deps.bzl", "main_deps", "main_exports", "main_resources", "test_deps")
load("@build_constants//:constants.bzl", "VERSION")
load("//jvm:build.bzl", "distribution")
load("@io_bazel_rules_kotlin//kotlin:android.bzl", "kt_android_library")
load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
load("@junit//junit5-jupiter-starter-bazel:junit5.bzl", "kt_jvm_junit5_test")
load("@rules_player//kotlin:lint.bzl", "lint")
load("@grab_bazel_common//tools/databinding:databinding.bzl", "kt_db_android_library")

kt_db_android_library(
    name = "player",
    srcs = glob(["src/main/java/**/*.kt"]),
    custom_package = "com.intuit.playerui.android",
    manifest = ":src/main/AndroidManifest.xml",
    resource_files = glob(["src/main/res/**"]),
    resources = main_resources,
    tags = ["maven_coordinates=com.intuit.playerui:android:aar:%s" % VERSION],
    visibility = ["//visibility:public"],
    exports = main_exports,
    deps = main_deps,
)

distribution(
    name = "player",
    lib_name = "player-databinding",
    maven_coordinates = "com.intuit.playerui:android:%s" % VERSION,
)

kt_jvm_junit5_test(
    name = "player-tests",
    srcs = glob(["src/test/java/**"]),
    associates = [":player-kotlin"],
    kotlinc_opts = "//jvm:test_options",
    test_package = "com.intuit.playerui.android",
    deps = [":player"] + test_deps,
)

lint(
    name = "player",
    srcs = glob(["src/**/*.kt"]),
    lint_config = "//jvm:lint_config",
)
