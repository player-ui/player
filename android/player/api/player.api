public final class com/intuit/player/android/AndroidPlayer : com/intuit/player/jvm/core/player/Player {
	public fun <init> (Ljava/util/List;)V
	public fun <init> ([Lcom/intuit/player/jvm/core/plugins/Plugin;)V
	public final fun commitPendingTransaction ()V
	public final fun getCachedStyledContext (Landroid/content/Context;Ljava/util/List;)Landroid/content/Context;
	public fun getHooks ()Lcom/intuit/player/android/AndroidPlayer$Hooks;
	public synthetic fun getHooks ()Lcom/intuit/player/jvm/core/player/Player$Hooks;
	public fun getLogger ()Lcom/intuit/player/jvm/core/logger/TapableLogger;
	public fun getPlugins ()Ljava/util/List;
	public fun getState ()Lcom/intuit/player/jvm/core/player/state/PlayerFlowState;
	public final fun onUpdate (Lkotlin/jvm/functions/Function2;)V
	public final fun recycle ()V
	public final fun registerAsset (Lkotlin/reflect/KClass;Ljava/lang/String;Lkotlin/jvm/functions/Function1;)V
	public final fun registerAsset (Lkotlin/reflect/KClass;Ljava/util/Map;Lkotlin/jvm/functions/Function1;)V
	public final fun registerPendingTransaction (ZLcom/intuit/player/android/AndroidPlayer$PendingTransaction;)V
	public static synthetic fun registerPendingTransaction$default (Lcom/intuit/player/android/AndroidPlayer;ZLcom/intuit/player/android/AndroidPlayer$PendingTransaction;ILjava/lang/Object;)V
	public fun release ()V
	public fun start (Ljava/lang/String;)Lcom/intuit/player/jvm/core/bridge/Completable;
}

public final class com/intuit/player/android/AndroidPlayer$Hooks : com/intuit/player/jvm/core/player/Player$Hooks {
	public final fun getContext ()Lcom/intuit/player/android/AndroidPlayer$Hooks$ContextHook;
	public fun getDataController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getExpressionEvaluator ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getFlowController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getState ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getView ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getViewController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
}

public final class com/intuit/player/android/AndroidPlayer$Hooks$ContextHook : com/intuit/hooks/SyncWaterfallHook {
	public fun <init> ()V
	public final fun call (Landroid/content/Context;)Landroid/content/Context;
}

public abstract interface class com/intuit/player/android/AndroidPlayer$PendingTransaction {
	public abstract fun commit ()V
}

public abstract interface class com/intuit/player/android/AndroidPlayerPlugin : com/intuit/player/jvm/core/plugins/Plugin {
	public abstract fun apply (Lcom/intuit/player/android/AndroidPlayer;)V
}

public final class com/intuit/player/android/AssetContext {
	public fun <init> (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;)V
	public fun <init> (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;)V
	public final fun component1 ()Landroid/content/Context;
	public final fun component2 ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun component3 ()Lcom/intuit/player/android/AndroidPlayer;
	public final fun component5 ()Ljava/lang/String;
	public final fun copy (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
	public static synthetic fun copy$default (Lcom/intuit/player/android/AssetContext;Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;ILjava/lang/Object;)Lcom/intuit/player/android/AssetContext;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getAsset ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun getContext ()Landroid/content/Context;
	public final fun getId ()Ljava/lang/String;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	public final fun getType ()Ljava/lang/String;
	public fun hashCode ()I
	public final fun postFixId (Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
	public fun toString ()Ljava/lang/String;
}

public final class com/intuit/player/android/AssetContextKt {
	public static final fun build (Lcom/intuit/player/android/AssetContext;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static final fun withContext (Lcom/intuit/player/android/AssetContext;Landroid/content/Context;)Lcom/intuit/player/android/AssetContext;
	public static final fun withStyles (Lcom/intuit/player/android/AssetContext;Ljava/util/List;)Lcom/intuit/player/android/AssetContext;
	public static final fun withStyles (Lcom/intuit/player/android/AssetContext;[Ljava/lang/Integer;)Lcom/intuit/player/android/AssetContext;
	public static final fun withTag (Lcom/intuit/player/android/AssetContext;Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
}

public final class com/intuit/player/android/BuildConfig {
	public static final field BUILD_TYPE Ljava/lang/String;
	public static final field DEBUG Z
	public static final field LIBRARY_PACKAGE_NAME Ljava/lang/String;
	public fun <init> ()V
}

public abstract class com/intuit/player/android/asset/DecodableAsset : com/intuit/player/android/asset/RenderableAsset {
	public fun <init> (Lcom/intuit/player/android/AssetContext;Lkotlinx/serialization/KSerializer;)V
	public final fun getData ()Ljava/lang/Object;
}

public abstract class com/intuit/player/android/asset/RenderableAsset : com/intuit/player/jvm/core/bridge/NodeWrapper {
	public fun <init> (Lcom/intuit/player/android/AssetContext;)V
	public final fun beacon (Ljava/lang/String;Ljava/lang/String;Lcom/intuit/player/jvm/core/asset/Asset;Ljava/lang/Object;)V
	public static synthetic fun beacon$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Ljava/lang/String;Lcom/intuit/player/jvm/core/asset/Asset;Ljava/lang/Object;ILjava/lang/Object;)V
	public final fun expand (Lcom/intuit/player/jvm/core/asset/AssetWrapper;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expand (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expand (Ljava/lang/String;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/asset/AssetWrapper;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expandList (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;)Ljava/util/List;
	public final fun expandList (Ljava/lang/String;Landroid/content/Context;)Ljava/util/List;
	public static synthetic fun expandList$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Ljava/util/List;
	public static synthetic fun expandList$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Ljava/util/List;
	public final fun getAsset ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun getAssetContext ()Lcom/intuit/player/android/AssetContext;
	public final fun getContext ()Landroid/content/Context;
	protected final fun getHydrationScope ()Lkotlinx/coroutines/CoroutineScope;
	public fun getNode ()Lcom/intuit/player/jvm/core/bridge/Node;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	protected final fun getUnitFunction (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;)Lcom/intuit/player/jvm/core/bridge/Invokable;
	protected abstract fun hydrate (Landroid/view/View;)V
	protected abstract fun initView ()Landroid/view/View;
	public final fun invalidateView ()V
	public final fun rehydrate ()V
	public final fun render (Landroid/content/Context;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/util/List;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/util/List;Ljava/lang/String;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;[Ljava/lang/Integer;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;[Ljava/lang/Integer;Ljava/lang/String;)Landroid/view/View;
	public final fun requireContext ()Landroid/content/Context;
}

public final class com/intuit/player/android/asset/RenderableAsset$Serializer : kotlinx/serialization/KSerializer {
	public fun <init> (Lcom/intuit/player/android/AndroidPlayer;)V
	public final fun conform (Lkotlin/reflect/KClass;)Lkotlinx/serialization/KSerializer;
	public fun deserialize (Lkotlinx/serialization/encoding/Decoder;)Lcom/intuit/player/android/asset/RenderableAsset;
	public synthetic fun deserialize (Lkotlinx/serialization/encoding/Decoder;)Ljava/lang/Object;
	public fun getDescriptor ()Lkotlinx/serialization/descriptors/SerialDescriptor;
	public fun serialize (Lkotlinx/serialization/encoding/Encoder;Lcom/intuit/player/android/asset/RenderableAsset;)Ljava/lang/Void;
	public synthetic fun serialize (Lkotlinx/serialization/encoding/Encoder;Ljava/lang/Object;)V
}

public abstract interface class com/intuit/player/android/asset/RenderableAsset$ViewportAsset {
}

public final class com/intuit/player/android/databinding/DefaultFallbackBinding : androidx/viewbinding/ViewBinding {
	public final field error Landroid/widget/TextView;
	public final field reset Landroid/widget/Button;
	public final field retry Landroid/widget/Button;
	public final field title Landroid/widget/TextView;
	public static fun bind (Landroid/view/View;)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
	public synthetic fun getRoot ()Landroid/view/View;
	public fun getRoot ()Landroidx/constraintlayout/widget/ConstraintLayout;
	public static fun inflate (Landroid/view/LayoutInflater;)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
	public static fun inflate (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Z)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
}

public final class com/intuit/player/android/databinding/FragmentPlayerBinding : androidx/viewbinding/ViewBinding {
	public final field playerCanvas Landroid/widget/FrameLayout;
	public static fun bind (Landroid/view/View;)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public synthetic fun getRoot ()Landroid/view/View;
	public fun getRoot ()Landroid/widget/ScrollView;
	public static fun inflate (Landroid/view/LayoutInflater;)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public static fun inflate (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Z)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
}

public final class com/intuit/player/android/extensions/IntoKt {
	public static final fun into (Landroid/view/View;Landroid/view/ViewGroup;)V
	public static final fun into (Landroid/view/View;Landroid/widget/FrameLayout;)V
	public static final fun into (Ljava/util/List;Landroid/view/ViewGroup;)V
	public static final fun transitionInto (Landroid/view/View;Landroid/widget/FrameLayout;Landroidx/transition/Transition;)V
}

public final class com/intuit/player/android/extensions/OverlayStylesKt {
	public static final fun overlayStyles (Landroid/content/Context;Ljava/util/List;Ljava/util/List;)Landroid/content/Context;
	public static final fun overlayStyles (Landroid/content/Context;[I)Landroid/content/Context;
	public static final fun overlayStyles (Landroid/content/Context;[ILjava/lang/Integer;)Landroid/content/Context;
	public static synthetic fun overlayStyles$default (Landroid/content/Context;Ljava/util/List;Ljava/util/List;ILjava/lang/Object;)Landroid/content/Context;
}

public abstract class com/intuit/player/android/lifecycle/ManagedPlayerState {
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Done : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Lcom/intuit/player/jvm/core/player/state/CompletedState;)V
	public final fun component1 ()Lcom/intuit/player/jvm/core/player/state/CompletedState;
	public final fun copy (Lcom/intuit/player/jvm/core/player/state/CompletedState;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;Lcom/intuit/player/jvm/core/player/state/CompletedState;ILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getCompletedState ()Lcom/intuit/player/jvm/core/player/state/CompletedState;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Error : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Ljava/lang/Exception;)V
	public final fun component1 ()Ljava/lang/Exception;
	public final fun copy (Ljava/lang/Exception;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;Ljava/lang/Exception;ILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getException ()Ljava/lang/Exception;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public abstract interface class com/intuit/player/android/lifecycle/ManagedPlayerState$Listener {
	public abstract fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public abstract fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public abstract fun onNotStarted ()V
	public abstract fun onPending ()V
	public abstract fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Listener$DefaultImpls {
	public static fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public static fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public static fun onNotStarted (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;)V
	public static fun onPending (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;)V
	public static fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$NotStarted : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public static final field INSTANCE Lcom/intuit/player/android/lifecycle/ManagedPlayerState$NotStarted;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Pending : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public static final field INSTANCE Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Pending;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Running : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Lcom/intuit/player/android/asset/RenderableAsset;Z)V
	public final fun component1 ()Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun component2 ()Z
	public final fun copy (Lcom/intuit/player/android/asset/RenderableAsset;Z)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;Lcom/intuit/player/android/asset/RenderableAsset;ZILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getAnimateViewTransition ()Z
	public final fun getAsset ()Lcom/intuit/player/android/asset/RenderableAsset;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public class com/intuit/player/android/lifecycle/PlayerViewModel : androidx/lifecycle/ViewModel, com/intuit/player/android/AndroidPlayerPlugin, com/intuit/player/jvm/core/plugins/RuntimePlugin {
	public fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;)V
	public fun apply (Lcom/intuit/player/android/AndroidPlayer;)V
	public fun apply (Lcom/intuit/player/jvm/core/bridge/runtime/Runtime;)V
	public final fun fail (Ljava/lang/Throwable;)V
	public final fun getBeacons ()Lkotlinx/coroutines/flow/SharedFlow;
	protected final fun getManager ()Lcom/intuit/player/jvm/core/managed/AsyncIterationManager;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	protected fun getPlugins ()Ljava/util/List;
	public final fun getState ()Lkotlinx/coroutines/flow/StateFlow;
	public fun onCleared ()V
	public final fun recycle ()V
	public final fun release ()V
	public final fun retry ()V
	public final fun start ()V
}

public final class com/intuit/player/android/lifecycle/PlayerViewModel$Factory : androidx/lifecycle/ViewModelProvider$Factory {
	public fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;Lkotlin/jvm/functions/Function1;)V
	public synthetic fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;Lkotlin/jvm/functions/Function1;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
	public fun create (Ljava/lang/Class;)Landroidx/lifecycle/ViewModel;
}

public final class com/intuit/player/android/lifecycle/PlayerViewModelKt {
	public static final fun fail (Lcom/intuit/player/android/lifecycle/PlayerViewModel;Ljava/lang/String;Ljava/lang/Throwable;)V
	public static synthetic fun fail$default (Lcom/intuit/player/android/lifecycle/PlayerViewModel;Ljava/lang/String;Ljava/lang/Throwable;ILjava/lang/Object;)V
}

public abstract class com/intuit/player/android/ui/PlayerFragment : androidx/fragment/app/Fragment, com/intuit/player/android/lifecycle/ManagedPlayerState$Listener {
	public fun <init> ()V
	public fun buildDoneView ()Landroid/view/View;
	public fun buildFallbackView (Ljava/lang/Exception;)Landroid/view/View;
	public fun buildLoadingView ()Landroid/view/View;
	public fun buildTransitionAnimation ()Landroidx/transition/Transition;
	protected final fun getBinding ()Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public abstract fun getPlayerViewModel ()Lcom/intuit/player/android/lifecycle/PlayerViewModel;
	protected fun handleAssetUpdate (Lcom/intuit/player/android/asset/RenderableAsset;Z)V
	public fun onCreateView (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Landroid/os/Bundle;)Landroid/view/View;
	public fun onDestroyView ()V
	public fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public fun onNotStarted ()V
	public fun onPending ()V
	public fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
	public final fun reset ()V
}

