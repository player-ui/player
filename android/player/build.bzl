load("@junit//junit5-jupiter-starter-bazel:junit5.bzl", "kt_jvm_junit5_test")
load("@rules_player//kotlin:lint.bzl", "lint")
load("//android/player:deps.bzl", "test_deps")

def kt_player_test(
        name,
        srcs = [],
        associates = [],
        deps = []):
    kt_jvm_junit5_test(
        name = name,
        srcs = srcs,
        associates = ["//android/player:player-associates"],
        kotlinc_opts = "//jvm:test_options",
        test_package = "com.intuit.playerui.android",
        deps = deps + [
            "//android/player:player",
            "//android/player/src/test/java/android/util:android_utils_stub",
            "//android/player/src/test/java/com/intuit/playerui/android/utils:android_test_utils",
            "//android/player/src/test/java/com/intuit/playerui/android/utils:android_log_util"] + test_deps,
    )

    lint(
        name = name,
        srcs = native.glob(["**/*.kt"]),
        lint_config = "//jvm:lint_config",
    )
