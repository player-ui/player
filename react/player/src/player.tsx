/* eslint-disable react/no-this-in-sfc */
import React from 'react';
import { SyncWaterfallHook, AsyncParallelHook } from 'tapable-ts';
import { Subscribe, useSubscribedState } from '@player-ui/react-subscribe';
import { Registry } from '@player-ui/partial-match-registry';
import type {
  CompletedState,
  PlayerPlugin,
  Flow,
  View,
} from '@player-ui/player';
import { Player } from '@player-ui/player';
import { ErrorBoundary } from 'react-error-boundary';
import type { AssetRegistryType } from './asset';
import { AssetContext } from './asset';
import { PlayerContext } from './utils';

import type { ReactPlayerProps } from './app';
import PlayerComp from './app';
import OnUpdatePlugin from './plugins/onupdate-plugin';

const WEB_PLAYER_VERSION = '__VERSION__';
const COMMIT = '__GIT_COMMIT__';

export interface DevtoolsGlobals {
  /** A global for a plugin to load to Player for devtools */
  __PLAYER_DEVTOOLS_PLUGIN?: {
    new (): ReactPlayerPlugin;
  };
}

export type DevtoolsWindow = typeof window & DevtoolsGlobals;

const _window: DevtoolsWindow | undefined =
  typeof window === 'undefined' ? undefined : window;

export interface ReactPlayerInfo {
  /** Version of the running player */
  playerVersion: string;

  /** Version of the running reactPlayer */
  reactPlayerVersion: string;

  /** Hash of the HEAD commit used to build the current player version */
  playerCommit: string;

  /** Hash of the HEAD commit used to build the current reactPlayer version */
  reactPlayerCommit: string;
}

export interface ReactPlayerPlugin extends Partial<PlayerPlugin> {
  /** The name of this plugin */
  name: string;

  /**
   * Attach listeners to the web-player instance
   */
  applyReact?: (reactPlayer: ReactPlayer) => void;
}

export interface ReactPlayerOptions {
  /** A headless player instance to use */
  player?: Player;

  /** A set of plugins to apply to this player */
  plugins?: Array<ReactPlayerPlugin>;

  /**
   * If the underlying reactPlayer.Component should use `React.Suspense` to trigger a loading state while waiting for content or content updates.
   * It requires that a `React.Suspense` component handler be somewhere in the `reactPlayer.Component` hierarchy.
   */
  suspend?: boolean;
}

export type ReactPlayerComponentProps = Record<string, unknown>;

/** A Player that renders UI through React */
export class ReactPlayer {
  public readonly options: ReactPlayerOptions;
  public readonly player: Player;
  public readonly assetRegistry: AssetRegistryType = new Registry();
  public readonly Component: React.ComponentType<ReactPlayerComponentProps>;
  public readonly hooks = {
    /**
     * A hook to create a React Component to be used for Player, regardless of the current flow state
     */
    webComponent: new SyncWaterfallHook<[React.ComponentType]>(),

    /**
     * A hook to create a React Component that's used to render a specific view.
     * It will be called for each view update from the core player.
     * Typically this will just be `Asset`
     */
    playerComponent: new SyncWaterfallHook<
      [React.ComponentType<ReactPlayerProps>]
    >(),

    /**
     * A hook to execute async tasks before the view resets to undefined
     */
    onBeforeViewReset: new AsyncParallelHook<[]>(),
  };

  private viewUpdateSubscription = new Subscribe<View>();
  private reactPlayerInfo: ReactPlayerInfo;

  constructor(options?: ReactPlayerOptions) {
    this.options = options ?? {};

    // Default the suspend option to `true` unless explicitly unset
    // Remove the suspend option in the next major
    if (!('suspend' in this.options)) {
      this.options.suspend = true;
    }

    const Devtools = _window?.__PLAYER_DEVTOOLS_PLUGIN;
    const onUpdatePlugin = new OnUpdatePlugin(
      this.viewUpdateSubscription.publish
    );

    const plugins = options?.plugins ?? [];

    if (Devtools) {
      plugins.push(new Devtools());
    }

    const playerPlugins = plugins.filter((p) =>
      Boolean(p.apply)
    ) as PlayerPlugin[];

    this.player = options?.player ?? new Player({ plugins: playerPlugins });

    plugins.forEach((plugin) => {
      if (plugin.applyReact) {
        plugin.applyReact(this);
      }
    });

    onUpdatePlugin.apply(this.player);

    this.Component = this.hooks.webComponent.call(this.createReactComp());
    this.reactPlayerInfo = {
      playerVersion: this.player.getVersion(),
      playerCommit: this.player.getCommit(),
      reactPlayerVersion: WEB_PLAYER_VERSION,
      reactPlayerCommit: COMMIT,
    };
  }

  /** Returns the current version of the underlying core Player */
  public getPlayerVersion(): string {
    return this.reactPlayerInfo.playerVersion;
  }

  /** Returns the git commit used to build this core Player version */
  public getPlayerCommit(): string {
    return this.reactPlayerInfo.playerCommit;
  }

  /** Find instance of [Plugin] that has been registered to the web player */
  public findPlugin<Plugin extends ReactPlayerPlugin>(
    symbol: symbol
  ): Plugin | undefined {
    return this.options.plugins?.find((el) => el.symbol === symbol) as Plugin;
  }

  /** Register and apply [Plugin] if one with the same symbol is not already registered. */
  public registerPlugin(plugin: ReactPlayerPlugin): void {
    if (!plugin.applyReact) return;

    plugin.applyReact(this);
    this.options.plugins?.push(plugin);
  }

  /** Returns the current version of the running React Player */
  public getReactPlayerVersion(): string {
    return this.reactPlayerInfo.reactPlayerVersion;
  }

  /** Returns the git commit used to build the React Player version */
  public getReactPlayerCommit(): string {
    return this.reactPlayerInfo.reactPlayerCommit;
  }

  private createReactComp(): React.ComponentType<ReactPlayerComponentProps> {
    const ActualPlayerComp = this.hooks.playerComponent.call(PlayerComp);

    /** the component to use to render Player */
    const ReactPlayerComponent = () => {
      const view = useSubscribedState<View>(this.viewUpdateSubscription);

      if (this.options.suspend) {
        this.viewUpdateSubscription.suspend();
      }

      return (
        <ErrorBoundary
          fallbackRender={() => null}
          onError={(err) => {
            const playerState = this.player.getState();

            if (playerState.status === 'in-progress') {
              playerState.fail(err);
            }
          }}
        >
          <PlayerContext.Provider value={{ player: this.player }}>
            <AssetContext.Provider
              value={{
                registry: this.assetRegistry,
              }}
            >
              {view && <ActualPlayerComp view={view} />}
            </AssetContext.Provider>
          </PlayerContext.Provider>
        </ErrorBoundary>
      );
    };

    return ReactPlayerComponent;
  }

  /**
   * Call this method to force the ReactPlayer to wait for the next view-update before performing the next render.
   * If the `suspense` option is set, this will suspend while an update is pending, otherwise nothing will be rendered.
   */
  public setWaitForNextViewUpdate() {
    // If the `suspend` option isn't set, then we need to reset immediately otherwise we risk flashing the old view while the new one is processing
    const shouldCallResetHook =
      this.options.suspend && this.hooks.onBeforeViewReset.isUsed();

    return this.viewUpdateSubscription.reset(
      shouldCallResetHook ? this.hooks.onBeforeViewReset.call() : undefined
    );
  }

  public start(flow: Flow): Promise<CompletedState> {
    this.setWaitForNextViewUpdate();

    return this.player.start(flow).finally(async () => {
      if (this.options?.suspend) {
        await this.setWaitForNextViewUpdate();
      }
    });
  }
}

// For compatibility
export const WebPlayer = ReactPlayer;
