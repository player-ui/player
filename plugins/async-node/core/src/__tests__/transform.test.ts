import { describe, it, expect } from "vitest";
import { asyncTransform } from "..";
import { Builder } from "@player-ui/player";

describe("asyncTransform", () => {
  const asset = Builder.asset({
    id: "2",
    type: "text",
    value: "chat message",
  });

  it("generates wrapper asset with asset", async () => {
    const transformedAsset = asyncTransform("1", "collection", asset);

    expect(transformedAsset).toMatchSnapshot();
  });

  it("generates wrapper asset with flatten is false", async () => {
    const transformedAsset = asyncTransform("1", "collection", asset, false);

    expect(transformedAsset?.children?.[0]?.value.values[1].flatten).toBe(
      false,
    );
  });
});
