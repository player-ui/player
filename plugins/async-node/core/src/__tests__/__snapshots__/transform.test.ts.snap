// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`asyncTransform > generates wrapper asset with asset 1`] = `
{
  "children": [
    {
      "path": [
        "values",
      ],
      "value": {
        "override": true,
        "parent": [Circular],
        "type": "multi-node",
        "values": [
          {
            "children": [
              {
                "path": [
                  "asset",
                ],
                "value": {
                  "parent": [Circular],
                  "type": "asset",
                  "value": {
                    "id": "2",
                    "type": "text",
                    "value": "chat message",
                  },
                },
              },
            ],
            "parent": [Circular],
            "type": "value",
            "value": undefined,
          },
          {
            "flatten": true,
            "id": "async-1",
            "parent": [Circular],
            "type": "async",
            "value": {
              "type": "value",
              "value": {
                "id": "async-1",
              },
            },
          },
        ],
      },
    },
  ],
  "type": "asset",
  "value": {
    "id": "collection-async-1",
    "type": "collection",
  },
}
`;
